packages/opencensus-core/src/metrics/gauges/derived-gauge.ts

```diff
@@ -0,0 +1,208 @@
+  private extractor: ValueExtractor;
+
+  private static readonly LABEL_VALUE = 'labelValue';
+  private static readonly LABEL_VALUES = 'labelValues';
+  private static readonly OBJECT = 'obj';
+  private static readonly NUMBER = 'number';
+  private static readonly FUNCTION = 'function';
+  private static readonly ERROR_MESSAGE_INVALID_SIZE =
+      'Label Keys and Label Values don\'t have same size';
+  private static readonly ERROR_MESSAGE_DUPLICATE_TIME_SERIES =
+      'A different time series with the same labels already exists.';
+  private static readonly ERROR_MESSAGE_UNKNOWN_INTERFACE =
+      'Unknown interface/object type';
+
+  /**
+   * Constructs a new DerivedGauge instance.
+   *
+   * @param {string} name The name of the metric.
+   * @param {string} description The description of the metric.
+   * @param {string} unit The unit of the metric.
+   * @param {MetricDescriptorType} type The type of metric.
+   * @param {LabelKey[]} labelKeys The list of the label keys.
+   */
+  constructor(
+      name: string, description: string, unit: string,
+      type: MetricDescriptorType, labelKeys: LabelKey[]) {
+    this.metricDescriptor = {name, description, unit, type, labelKeys};
+    this.labelKeysLength = labelKeys.length;
+  }
+
+  // Checks if the specified collection is a LengthAttributeInterface.
+  // tslint:disable-next-line:no-any
+  protected static isLengthAttributeInterface(obj: any):
+      obj is LengthAttributeInterface {
+    return obj && typeof obj.length === DerivedGauge.NUMBER;
+  }
+
+  // Checks if the specified collection is a LengthMethodInterface.
+  // tslint:disable-next-line:no-any
+  protected static isLengthMethodInterface(obj: any):
+      obj is LengthMethodInterface {
+    return obj && typeof obj.length === DerivedGauge.FUNCTION;
+  }
+
+  // Checks if the specified collection is a SizeAttributeInterface.
+  // tslint:disable-next-line:no-any
+  protected static isSizeAttributeInterface(obj: any):
+      obj is SizeAttributeInterface {
+    return obj && typeof obj.size === DerivedGauge.NUMBER;
+  }
+
+  // Checks if the specified collection is a SizeMethodInterface.
+  // tslint:disable-next-line:no-any
+  protected static isSizeMethodInterface(obj: any): obj is SizeMethodInterface {
+    return obj && typeof obj.size === DerivedGauge.FUNCTION;
+  }
+
+  // Checks if the specified callbackFn is a ToValueInterface.
+  // tslint:disable-next-line:no-any
+  protected static isToValueInterface(obj: any): obj is ToValueInterface {
+    return obj && typeof obj.getValue === DerivedGauge.FUNCTION;
+  }
+
+  /**
+   * Creates a TimeSeries. The value of a single point in the TimeSeries is
+   * observed from a obj. The ValueExtractor is invoked whenever
+   * metrics are collected, meaning the reported value is up-to-date.
+   *
+   * @param {LabelValue[]} labelValues The list of the label values.
+   * @param obj The obj to get the size / length / value from.
```

Comment:
nit: You can also mention the order if multiple options are available, e.g value -> length -> size.

Reply:
Done.